clang/lib/Sema/SemaChecking.cpp

```diff
@@ -1938,11 +1938,25 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
     if (Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)) {
       assert(Context.getAuxTargetInfo() &&
              "Aux Target Builtin, but not an aux target?");
-
+      /*
+      'CheckTSBuiltinFunctionCall' function checks for an aux-target builtin and also validates it 
+      against the correct target. */
       if (CheckTSBuiltinFunctionCall(
               Context.getAuxTargetInfo()->getTriple().getArch(),
-              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall))
-        return ExprError();
+              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall)) {
+                /*At this point, we know that the builtin is an aux-builtin and also checked.
+                  AUX builtins are not allowed to appear inside device code.
+                  To handle this, we use the notion of "deferred diagnostics", where we are compiling
+                  for the device, but we don't know that this function will be codegen'ed
+                  for device yet. So we create a diagnostic which is emitted if and when we
+                  realize that the function will be codegen'ed
+                */
+                if (getLangOpts().SYCLIsDevice) {
```

Comment:
This isn't right,tis will only diagnose when the builtin has some sort of error. (CheckTSBuiltinFunction call returns true on failure).  We STILL want to report those and let that be an expr error.  Otherwise we want to diagnose like youre doing below.

Reply:
I don't understand your comment. 
If `Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)` is true, then the builtinID is an AUX builtin.
Then we call CheckTSBuiltinFunctionCall with the target architecture of the AUX builtin and the builtinID. If this is true, it means the AUX builtin does belong to the target specified. Right?
Then I check if we are inside device code or not, since we should only identify aux bultins inside device code. So I used - `if (getLangOpts().SYCLIsDevice)` which will be true, if we are inside device code?
>>(CheckTSBuiltinFunction call returns true on failure)
>>tis will only diagnose when the builtin has some sort of error.
Can you clarify what these 2 statements mean?

Reply:
> I don't understand your comment.
> If `Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)` is true, then the builtinID is an AUX builtin.
> Then we call CheckTSBuiltinFunctionCall with the target architecture of the AUX builtin and the builtinID. If this is true, it means the AUX builtin does belong to the target specified. Right?

isAuxBuiltinID means that the target arch of the aux builtin is the aux target.  That part is already calculated.  The CheckTSBuiltinFunctionCall validates the parameters to said function call..  If it is 'true' this check fails. 

> Then I check if we are inside device code or not, since we should only identify aux bultins inside device code. So I used - `if (getLangOpts().SYCLIsDevice)` which will be true, if we are inside device code?
Correct.
> 
> > > (CheckTSBuiltinFunction call returns true on failure)
> > > tis will only diagnose when the builtin has some sort of error.
> > > Can you clarify what these 2 statements mean?

Since the function is checking the parameters to the builtin, it does 1 of 2 things:  Returns 'false' meaning there was no errors.  OR  Emits an error (like, bad parameter value), then returns true.

Reply:
Usually, when bool functions return true, it means something worked correctly and false means something went wrong.
This reverse logic is confusing.

Reply:
The CFE typically has 'true' mean "had an error", its just the convention that was done when they started.

Reply:
The function name could be `ValidateBuiltinFuncParams `and it could return true is no errors and false if there is some error.

Reply:
Thats just not the convention of the CFE.  Also, "Check" is a very common word for exactly this throughout the sema layer.  Doing it differently would be more confusing to those who are familiar with the code.